homeassistant/components/qwikswitch.py

```diff
@@ -76,127 +79,98 @@ def is_on(self):
+    hass.data[DOMAIN] = qsusb
 
-    # Discover all devices in QSUSB
-    yield from qsusb.update_from_devices()
-    hass.data[DOMAIN]['switch'] = []
-    hass.data[DOMAIN]['light'] = []
+    _new = {'switch': [], 'light': [], 'sensor': sensors}
     for _id, item in qsusb.devices:
-        if (item[QS_TYPE] == QSType.relay and
-                item[QSDATA][QS_NAME].lower().endswith(' switch')):
-            item[QSDATA][QS_NAME] = item[QSDATA][QS_NAME][:-7]  # Remove switch
-            new_dev = QSSwitch(_id, qsusb)
-            hass.data[DOMAIN]['switch'].append(new_dev)
+        if _id in switches:
+            if item[QS_TYPE] != QSType.relay:
+                _LOGGER.warning(
+                    "You specified a switch that is not a relay %s", _id)
+                continue
+            _new['switch'].append(_id)
         elif item[QS_TYPE] in [QSType.relay, QSType.dimmer]:
-            new_dev = QSLight(_id, qsusb)
-            hass.data[DOMAIN]['light'].append(new_dev)
+            _new['light'].append(_id)
         else:
             _LOGGER.warning("Ignored unknown QSUSB device: %s", item)
             continue
-        hass.data[DOMAIN][_id] = new_dev
 
     # Load platforms
-    for comp_name in ('switch', 'light'):
-        if hass.data[DOMAIN][comp_name]:
-            load_platform(hass, comp_name, 'qwikswitch', {}, config)
+    for comp_name in _new:
+        if _new[comp_name]:
+            load_platform(
+                hass, comp_name, DOMAIN, {DOMAIN: _new[comp_name]}, config)
 
     def callback_qs_listen(item):
         """Typically a button press or update signal."""
-        if qsusb is None:  # Shutting down
-            return
-
         # If button pressed, fire a hass event
-        if item.get(QS_CMD, '') in cmd_buttons and QS_ID in item:
-            hass.bus.async_fire('qwikswitch.button.{}'.format(item[QS_ID]))
-            return
+        if QS_ID in item:
+            if item.get(QS_CMD, '') in cmd_buttons:
+                hass.bus.async_fire(
+                    'qwikswitch.button.{}'.format(item[QS_ID]), item)
+                return
+
+            if item[QS_ID] not in qsusb.devices:
+                # Not a standard device in, component can handle packet
+                # i.e. sensors
+                hass.helpers.dispatcher.async_dispatcher_send(
+                    item[QS_ID], item)
 
         # Update all ha_objects
         hass.async_add_job(qsusb.update_from_devices)
 
     @callback
-    def async_start(event):  # pylint: disable=unused-argument
+    def async_start(_):
         """Start listening."""
         hass.async_add_job(qsusb.listen, callback_qs_listen)
 
     hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, async_start)
 
+    @callback
+    def async_stop(_):
+        """Stop the listener queue and clean up."""
+        hass.data[DOMAIN].stop()
+        _LOGGER.info("Waiting for long poll to QSUSB to time out (max 30sec)")
```

Comment:
This isn't blocking right?

Reply:
`stop` is not bloking, it only sets a _running varialble False so that I will exit my listen loop. Since it is a long poll it can potentiall take up to 30seconds for asyncio.Timeout - see [here](https://github.com/kellerza/pyqwikswitch/blob/master/pyqwikswitch/async_.py#L78)

Maybe ther is a way to cancel such a task, but since it is only on shutdown I am not too worried

Reply:
If it's a asyncio task or future, you can call its cancel method. If you have written a coroutine and wrapped that in a task, you can except `CancelledError` to clean up.